torch/csrc/utils/tensor_new.cpp

```diff
@@ -268,6 +281,9 @@ Tensor internal_new_from_data(
     auto device = device_opt.has_value() ? *device_opt : at::Device(computeDeviceType(type_id));
     AutoNoGIL no_gil;
     maybe_initialize_cuda(device);
+#ifdef BUILD_NAMEDTENSOR
+    at::namedinference::propagate_names(tensor, names);
+#endif
```

Comment:
seems like this code should be refactored a bit.

E.g. as far as I can tell the body of the two USE_NUMPY conditions are nearly identical, so if you refactor you'd need fewer `#ifdef BUILD_NAMEDTENSOR` blocks, which would make the code more readable. Ideally you'd only need one. I think with better code re-use this code would be more readable regardless of your change.

Reply:
If I refactored this into a function, that function would require a guard on one of its arguments based on `#ifdef BUILD_NAMEDTENSOR` because it would take in a names argument.

We'd still have the same number of `#ifdef BUILD_NAMEDTENSOR` blocks (actually we'd have ~2 more because of the function declaration and definition), so this would not clean up the code.

I agree that this code could use a refactor, however. I can file a task and take care of it when I delete the `build_namedtensor` flag (that is slated to happen sometime hopefully next week).

@nairbv Did you have any other concerns on the code?

Reply:
I don't (necessarily) mean separate functions.

e.g. if we just wrote `if (PyObject_HasAttrString(data, "__cuda_array_interface__") || PyArray_Check(data))` above instead of having two separate conditions, the only thing that differs inside the if block is the call to get the tensor. The repeated code doesn't necessarily need to be moved to a helper function to be combined.

I guess this can be refactored some other time though.

Reply:
> e.g. if we just wrote if (PyObject_HasAttrString(data, "__cuda_array_interface__") || PyArray_Check(data)) above instead of having two separate conditions, the only thing that differs inside the if block is the call to get the tensor. The repeated code doesn't necessarily need to be moved to a helper function to be combined.

I agree that it would be marginally cleaner, but not to the extent that we should block this PR on. A full refactor of the code that would make the logic in this function a lot clearer, in my opinion, would split up each of the codepaths into distinct cases (in the form of helper functions): `build_tensor_from_cuda_array`, `build_tensor_from_numpy`, `build_tensor_from_pylist`, `build_tensor_from_tensor`. `build_tensor_from_cuda_array` would share a lot of common code with `build_tensor_from_numpy`, and all four of these functions would share a lot of common code in the place where we coerce the input to a device/dtype/names. However, if I were to do this right now, each function would be littered with `#ifdef BUILD_NAMEDTENSOR` blocks.